# Sold Data EDA (Week2) 
Using the dataset filtered "PropertyType == Residential"  
Author : Yoora Choi  

In [55]:
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

In [56]:
# Import listed data (2024.01-2026.03)
df = pd.read_csv("/Users/yoorachoi/Python/IDX/Sold/data/CRMLSSold_merge.csv", low_memory=False)

## 1.Data overview

In [57]:
# Check data structure
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 397210 entries, 0 to 397209
Data columns (total 84 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   BuyerAgentAOR                 348153 non-null  object 
 1   ListAgentAOR                  351024 non-null  object 
 2   Flooring                      254779 non-null  object 
 3   ViewYN                        363391 non-null  object 
 4   WaterfrontYN                  248 non-null     object 
 5   BasementYN                    7770 non-null    object 
 6   PoolPrivateYN                 362794 non-null  object 
 7   OriginalListPrice             396496 non-null  float64
 8   ListingKey                    397210 non-null  int64  
 9   ListAgentEmail                368591 non-null  object 
 10  CloseDate                     397210 non-null  object 
 11  ClosePrice                    397208 non-null  float64
 12  ListAgentFirstName            394211 non-nul

In [58]:
# check the data structure
df.head(5)

,BuyerAgentAOR,ListAgentAOR,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,CoListAgentFirstName,CoListAgentLastName,BuyerAgentMlsId,BuyerAgentFirstName,BuyerAgentLastName,FireplacesTotal,AssociationFeeFrequency,AboveGradeFinishedArea,ListingKeyNumeric,MLSAreaMajor,TaxAnnualAmount,CountyOrParish,MlsStatus,ElementarySchool,AttachedGarageYN,ParkingTotal,BuilderName,PropertySubType,LotSizeAcres,SubdivisionName,BuyerOfficeAOR,YearBuilt,StreetNumberNumeric,ListingId,BathroomsTotalInteger,City,TaxYear,BuildingAreaTotal,BedroomsTotal,ContractStatusChangeDate,ElementarySchoolDistrict,CoBuyerAgentFirstName,PurchaseContractDate,ListingContractDate,BelowGradeFinishedArea,BusinessType,StateOrProvince,CoveredSpaces,MiddleOrJuniorSchool,FireplaceYN,Stories,HighSchool,Levels,LotSizeDimensions,LotSizeArea,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,OriginatingSystemName,OriginatingSystemSubName,BuyerAgencyCompensationType,BuyerAgencyCompensation,latfilled,lonfilled
0,Mlslistings,Mlslistings,"Carpet,Tile,Wood",True,NaN,NaN,False,499000.0,551985747,jwachter@cbnorcal.com,2024-01-26,240000.0,Joan,Wachter,NaN,NaN,1 Baldwin Avenue 411,Residential,1140.0,295000.0,777,Coldwell Banker Realty,Bay Area Senior Services,NaN,Joan Wachter,NaN,NaN,ML5090968,Scott,Withrow,NaN,Monthly,NaN,551985747,699 - Not Defined,NaN,San Mateo,Closed,NaN,False,1.0,NaN,Condominium,NaN,NaN,Mlslistings,1988.0,1.0,ML81865679,2.0,San Mateo,NaN,NaN,2.0,2024-01-26,NaN,NaN,2023-11-22,2021-10-06,NaN,NaN,CA,NaN,NaN,True,NaN,NaN,NaN,NaN,NaN,NaN,False,1.0,Other,94401,6472.0,NaN,NaN,CRMLS,CRMLS_MLSL,NaN,NaN,NaN,NaN
1,SanDiego,SanDiego,NaN,False,NaN,NaN,False,759900.0,522107581,mdarwich12@gmail.com,2024-01-05,815000.0,Michael,Darwich,NaN,NaN,2811 C Avenue,Residential,1974.0,759900.0,33,Berkshire Hathaway HomeServices California Pro...,FOSTER HAMILTON Real Estate,NaN,Michael Darwich,NaN,NaN,SAND-606727,Charles,Street,NaN,NaN,NaN,522107581,91950 - National City,NaN,San Diego,Closed,NaN,True,2.0,NaN,SingleFamilyResidence,NaN,National City,SanDiego,2023.0,2811.0,210014555,4.0,National City,NaN,NaN,4.0,2024-01-05,NaN,NaN,2021-06-30,2021-03-08,NaN,NaN,CA,NaN,NaN,False,2.0,NaN,Two,NaN,NaN,NaN,False,2.0,NaN,91950,NaN,NaN,NaN,CRMLS,CRMLS_SAND,NaN,NaN,NaN,NaN
2,SanDiego,SanDiego,NaN,False,NaN,NaN,False,739900.0,510919001,mdarwich12@gmail.com,2024-01-05,810000.0,Michael,Darwich,NaN,NaN,2812 C Avenue,Residential,1974.0,770000.0,228,Berkshire Hathaway HomeServices California Pro...,FOSTER HAMILTON Real Estate,NaN,Michael Darwich,NaN,NaN,SAND-606727,Charles,Street,NaN,NaN,NaN,510919001,91950 - National City,NaN,San Diego,Closed,NaN,True,2.0,NaN,SingleFamilyResidence,NaN,National City,SanDiego,2023.0,2812.0,210008330,4.0,National City,NaN,NaN,4.0,2024-01-05,NaN,NaN,2021-11-18,2021-03-08,NaN,NaN,CA,NaN,NaN,False,2.0,NaN,Two,NaN,NaN,NaN,False,2.0,NaN,91950,NaN,NaN,NaN,CRMLS,CRMLS_SAND,NaN,NaN,NaN,NaN
3,Mlslistings,Mlslistings,NaN,False,NaN,NaN,NaN,NaN,1079166779,davidmartz@compass.com,2024-01-30,858000.0,David,Martz,NaN,NaN,2199 N Berne Drive,Residential,1995.0,858000.0,0,Compass,NaN,NaN,David Martz,NaN,NaN,ML921886,Steven,Hill,NaN,NaN,NaN,1079166779,NaN,NaN,Riverside,Closed,Vista Del Monte,False,2.0,NaN,SingleFamilyResidence,0.3100,NaN,NaN,1959.0,2199.0,ML81975626,3.0,Palm Springs,NaN,NaN,0.0,2024-01-30,NaN,NaN,2024-08-05,2024-01-30,NaN,NaN,CA,NaN,Raymond Cree,False,NaN,NaN,NaN,NaN,13504.0,NaN,False,2.0,Palm Springs Unified,92262,NaN,13504.0,NaN,CRMLS,CRMLS_MLSL,NaN,NaN,NaN,NaN
4,Southland,Southland,NaN,False,NaN,NaN,False,1890500.0,1075037759,karen.klein@theagencyre.com,2024-01-29,1890500.0,Karen,Klein,34.164376,-118.558749,19524 Winifred Street,Residentia

In [59]:
# Summary statistics for numerical variable
df.describe()

,OriginalListPrice,ListingKey,ClosePrice,Latitude,Longitude,LivingArea,ListPrice,DaysOnMarket,FireplacesTotal,AboveGradeFinishedArea,ListingKeyNumeric,TaxAnnualAmount,ParkingTotal,LotSizeAcres,YearBuilt,StreetNumberNumeric,BathroomsTotalInteger,TaxYear,BuildingAreaTotal,BedroomsTotal,ElementarySchoolDistrict,BelowGradeFinishedArea,BusinessType,CoveredSpaces,Stories,LotSizeArea,MainLevelBedrooms,GarageSpaces,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,BuyerAgencyCompensation
count,3.964960e+05,3.972100e+05,3.972080e+05,381396.000000,381396.000000,3.969820e+05,3.972100e+05,397210.000000,0.0,0.0,3.972100e+05,0.0,3.967840e+05,3.658920e+05,396854.000000,3.967680e+05,397141.000000,0.0,27751.000000,397199.000000,0.0,2289.000000,0.0,0.0,335756.000000,3.663930e+05,230561.000000,380187.000000,306314.000000,3.662700e+05,0.0,46125.000000
mean,1.224871e+06,1.095857e+09,1.186358e+06,34.652719,-118.569272,1.904399e+03,1.138657e+06,37.323796,NaN,NaN,1.095857e+09,NaN,9.331131e+00,6.838508e+01,1978.562837,1.044145e+04,2.533148,NaN,2118.842312,3.200237,NaN,98.043342,NaN,NaN,1.364270,3.839939e+04,2.059151,1.858525,234.812115,3.349959e+05,NaN,221.929675
std,6.783012e+06,2.990261e+07,5.982507e+06,1.707120,2.957121,2.703114e+04,1.353876e+06,53.514134,NaN,NaN,2.990261e+07,NaN,4.117696e+03,1.633434e+04,26.268146,2.289134e+05,1.136101,NaN,1705.364856,1.067473,NaN,737.200389,NaN,NaN,0.481226,2.140740e+06,7.244967,3.708293,1773.112598,1.631441e+07,NaN,2590.406598
min,0.000000e+00,4.216785e+08,0.000000e+00,-117.472493,-177.646696,0.000000e+00,5.250000e+02,-288.000000,NaN,NaN,4.216785e+08,NaN,-1.430000e+02,0.000000e+00,1776.000000,0.000000e+00,0.000000,NaN,0.000000,0.000000,NaN,0.000000,NaN,NaN,1.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000e+00,NaN,0.000000
25%,5.850000e+05,1.075063e+09,5.750000e+05,33.736511,-118.880418,1.247000e+03,5.751750e+05,8.000000,NaN,NaN,1.075063e+09,NaN,2.000000e+00,1.200000e-01,1960.000000,1.006000e+03,2.000000,NaN,1270.000000,3.000000,NaN,0.000000,NaN,NaN,1.000000,5.000000e+03,1.000000,2.000000,0.000000,5.227000e+03,NaN,2.000000
50%,8.250000e+05,1.093934e+09,8.200000e+05,34.046786,-118.014369,1.641000e+03,8.150000e+05,19.000000,NaN,NaN,1.093934e+09,NaN,2.000000e+00,1.663000e-01,1979.000000,3.300000e+03,2.000000,NaN,1714.000000,3.000000,NaN,0.000000,NaN,NaN,1.000000,7.016000e+03,2.000000,2.000000,95.000000,7.242000e+03,NaN,2.500000
75%,1.299000e+06,1.115062e+09,1.300000e+06,34.552973,-117.267968,2.217000e+03,1.295000e+06,48.000000,NaN,NaN,1.115062e+09,NaN,3.000000e+00,2.723000e-01,1999.000000,1.202900e+04,3.000000,NaN,2459.000000,4.000000,NaN,0.000000,NaN,NaN,2.000000,1.089000e+04,3.000000,2.000000,368.507500,1.182200e+04,NaN,2.500000
max,1.390000e+09,1.156473e+09,9.895000e+08,56.130366,329.000000,1.702132e+07,1.375000e+08,12430.000000,NaN,NaN,1.156473e+09,NaN,2.593669e+06,7.810698e+06,2026.000000,7.814782e+07,175.000000,NaN,123764.000000,45.000000,NaN,14531.000000,NaN,NaN,2.000000,9.187423e+08,3333.000000,600.000000,750000.000000,2.208318e+09,NaN,100000.000000


In [60]:
# Summary statistics for non-numerical variable
df.describe(include = 'object')

,BuyerAgentAOR,ListAgentAOR,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,ListAgentEmail,CloseDate,ListAgentFirstName,ListAgentLastName,UnparsedAddress,PropertyType,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,CoListAgentFirstName,CoListAgentLastName,BuyerAgentMlsId,BuyerAgentFirstName,BuyerAgentLastName,AssociationFeeFrequency,MLSAreaMajor,CountyOrParish,MlsStatus,ElementarySchool,AttachedGarageYN,BuilderName,PropertySubType,SubdivisionName,BuyerOfficeAOR,ListingId,City,ContractStatusChangeDate,CoBuyerAgentFirstName,PurchaseContractDate,ListingContractDate,StateOrProvince,MiddleOrJuniorSchool,FireplaceYN,HighSchool,Levels,LotSizeDimensions,NewConstructionYN,HighSchoolDistrict,PostalCode,OriginatingSystemName,OriginatingSystemSubName,BuyerAgencyCompensationType,latfilled,lonfilled
count,348153,351024,254779,363391,248,7770,362794,368591,397210,394211,397170,396851,397210,397210,390677,93996,397125,90906,91141,396708,395256,397043,165369,344153,397210,397210,52955,336596,19511,396434,147582,375798,397210,396900,396621,35986,397016,397209,397210,53306,396894,69210,358662,19301,367789,288500,397208,38859,38859,46136,63884,63884
unique,63,59,326,2,1,1,2,79446,818,15250,34187,386352,1,18397,20954,5429,76731,6566,13640,113685,18968,39477,4,1088,63,1,1394,2,3281,20,15776,69,396878,1123,818,4309,1135,1417,15,660,2,520,19,9652,2,444,3577,1,10,3,2,2
top,OrangeCounty,OrangeCounty,"Carpet,Tile",True,True,True,False,homes@opendoor.com,2024-06-28,Michael,Lee,5508 Avenida Maravillas,Residential,Compass,Compass,Compass,RANDY ANDERSON,Michael,Myatt,NONMEMBER,General,NONMEMBER,Monthly,699 - Not Defined,Los Angeles,Closed,Other,True,Lennar,SingleFamilyResidence,Other,OrangeCounty,OR24107765,San Diego,2024-06-28,Michael,2024-04-24,2025-05-01,CA,Other,True,Other,One,50x135,False,Other,92253,CRMLS,CRMLS_CRM,Item1,True,True
freq,29350,29573,28532,220591,248,7770,313552,1068,1434,4828,1840,4,397210,27780,25890,10158,988,1333,891,7537,6172,6299,155294,41907,98730,397210,9117,266855,1819,297065,4041,31665,3,17851,1434,527,874,1152,397182,6782,262171,2719,211657,334,352793,30545,2565,38859,21482,45479,33427,33427


## 2. Missing data check(null)

In [61]:
# null count
null_count = df.isnull().sum()

# null percentage (%)
null_percent = (df.isnull().sum() / len(df)) * 100

# data frame
null_summary = pd.DataFrame({
    'null_count': null_count,
    'null_percent': null_percent
}).sort_values(by='null_percent', ascending=False)

null_summary['null_percent'] = null_summary['null_percent'].round(2)
null_summary

,null_count,null_percent
CoveredSpaces,397210,100.00
MiddleOrJuniorSchoolDistrict,397210,100.00
AboveGradeFinishedArea,397210,100.00
FireplacesTotal,397210,100.00
TaxYear,397210,100.00
ElementarySchoolDistrict,397210,100.00
BusinessType,397210,100.00
TaxAnnualAmount,397210,100.00
WaterfrontYN,396962,99.94
BelowGradeFinishedArea,394921,99.42


## 3. Check duplicates

### 3.1 duplicated columns

In [62]:
# check duplicate columns
pairs = []

for col in df.columns:
    if '.' in col:
        base_col = col.split('.')[0]
        if base_col in df.columns:
            pairs.append((base_col, col))

pairs

[]

### 3.2 duplicated rows

In [63]:
# check completely duplicated rows
#df[df.duplicated(keep=False)]

In [64]:
df.duplicated().sum()

np.int64(48)

In [65]:
# Key MLS Fields - Numerical summary
key_num_cols = ['ListPrice', 'ClosePrice', 'LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger', 'DaysOnMarket', 'YearBuilt', 'OriginalListPrice','LotSizeAcres']
df[key_num_cols].describe().round(2)

,ListPrice,ClosePrice,LivingArea,BedroomsTotal,BathroomsTotalInteger,DaysOnMarket,YearBuilt,OriginalListPrice,LotSizeAcres
count,3.972100e+05,3.972080e+05,396982.00,397199.00,397141.00,397210.00,396854.00,3.964960e+05,365892.00
mean,1.138657e+06,1.186358e+06,1904.40,3.20,2.53,37.32,1978.56,1.224871e+06,68.39
std,1.353876e+06,5.982507e+06,27031.14,1.07,1.14,53.51,26.27,6.783012e+06,16334.34
min,5.250000e+02,0.000000e+00,0.00,0.00,0.00,-288.00,1776.00,0.000000e+00,0.00
25%,5.751750e+05,5.750000e+05,1247.00,3.00,2.00,8.00,1960.00,5.850000e+05,0.12
50%,8.150000e+05,8.200000e+05,1641.00,3.00,2.00,19.00,1979.00,8.250000e+05,0.17
75%,1.295000e+06,1.300000e+06,2217.00,4.00,3.00,48.00,1999.00,1.299000e+06,0.27
max,1.375000e+08,9.895000e+08,17021321.00,45.00,175.00,12430.00,2026.00,1.390000e+09,7810698.36


# Sold Data EDA — Week 2 Summary
**Dataset:** CRMLS Sold Listings (Jan 2024 – Mar 2026)  
**Filter Applied:** `PropertyType` == Residential   
**Author:** Yoora Choi

---

## 1. Dataset Overview

| Item | Detail |
|------|--------|
| Total Rows | 397,210 |
| Total Columns | 84 |
| Data Period | January 2024 – March 2026 |
| Data Types | float64: 29 / int64: 3 / object: 52 |

---

## 2. Missing Value Analysis

### 100% Missing — Drop Immediately (8 columns)

| Column | Missing Rate |
|--------|-------------|
| `CoveredSpaces` | 100% |
| `MiddleOrJuniorSchoolDistrict` | 100% |
| `AboveGradeFinishedArea` | 100% |
| `FireplacesTotal` | 100% |
| `TaxYear` | 100% |
| `ElementarySchoolDistrict` | 100% |
| `BusinessType` | 100% |
| `TaxAnnualAmount` | 100% |

---

### 90–100% Missing (9 columns)

| Column | Missing Rate |
|--------|-------------|
| `WaterfrontYN` | 99.94% |
| `BelowGradeFinishedArea` | 99.42% |
| `BasementYN` | 98.04% |
| `LotSizeDimensions` | 95.14% |
| `BuilderName` | 95.09% |
| `BuildingAreaTotal` | 93.01% |
| `CoBuyerAgentFirstName` | 90.94% |
| `OriginatingSystemName` | 90.22% |
| `OriginatingSystemSubName` | 90.22% |

---

### 70–90% Missing

| Column | Missing Rate |
|--------|-------------|
| `BuyerAgencyCompensation` | 88.39% |
| `BuyerAgencyCompensationType` | 88.38% |
| `ElementarySchool` | 86.67% |
| `MiddleOrJuniorSchool` | 86.58% |
| `latfilled` | 83.92% |
| `lonfilled` | 83.92% |
| `HighSchool` | 82.58% |
| `CoListAgentFirstName` | 77.11% |
| `CoListAgentLastName` | 77.05% |
| `CoListOfficeName` | 76.34% |

---

### 10–70% Missing

| Column | Missing Rate |
|--------|-------------|
| `SubdivisionName` | 62.85% |
| `AssociationFeeFrequency` | 58.37% |
| `MainLevelBedrooms` | 41.95% |
| `Flooring` | 35.86% |
| `HighSchoolDistrict` | 27.37% |
| `AssociationFee` | 22.88% |
| `Stories` | 15.47% |
| `AttachedGarageYN` | 15.26% |
| `MLSAreaMajor` | 13.36% |
| `BuyerAgentAOR` | 12.35% |
| `ListAgentAOR` | 11.63% |

---

### 0–10% Missing

| Column | Missing Rate |
|--------|-------------|
| `Levels` | 9.70% |
| `PoolPrivateYN` | 8.66% |
| `ViewYN` | 8.51% |
| `LotSizeAcres` | 7.88% |
| `LotSizeSquareFeet` | 7.79% |
| `LotSizeArea` | 7.76% |
| `NewConstructionYN` | 7.41% |
| `ListAgentEmail` | 7.21% |
| `BuyerOfficeAOR` | 5.39% |
| `GarageSpaces` | 4.29% |
| `Latitude / Longitude` | 3.98% |
| `BuyerOfficeName` | 1.64% |
| `ListAgentFirstName` | 0.76% |
| `BuyerAgentFirstName` | 0.49% |
| `PropertySubType` | 0.20% |
| `OriginalListPrice` | 0.18% |
| `YearBuilt` | 0.09% |
| `LivingArea` | 0.06% |
| `BathroomsTotalInteger` | 0.02% |
| `BedroomsTotal` | 0.00% |

### No Missing Values

`CloseDate`, `ClosePrice`, `ListPrice`, `DaysOnMarket`, `ListingKey`, `ListingKeyNumeric`, `ListOfficeName`, `MlsStatus`, `PropertyType`, `StateOrProvince`, `CountyOrParish`, `ListingId`

---

## 3. Duplicate Analysis

### Duplicate Rows

> **48 fully duplicate rows found.** 

---

### Duplicate Column Pairs

> **No duplicate column pairs found.** Unlike the listing dataset, this dataset has no `.1` suffix columns — cleaner structure overall.

---

## 4. Key Numeric Field Distributions



In [66]:

key_num_cols = ['OriginalListPrice', 'ListPrice', 'ClosePrice', 'LotSizeAcres',
                'LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger',
                'DaysOnMarket', 'YearBuilt']
df[key_num_cols].describe().round(2)



,OriginalListPrice,ListPrice,ClosePrice,LotSizeAcres,LivingArea,BedroomsTotal,BathroomsTotalInteger,DaysOnMarket,YearBuilt
count,3.964960e+05,3.972100e+05,3.972080e+05,365892.00,396982.00,397199.00,397141.00,397210.00,396854.00
mean,1.224871e+06,1.138657e+06,1.186358e+06,68.39,1904.40,3.20,2.53,37.32,1978.56
std,6.783012e+06,1.353876e+06,5.982507e+06,16334.34,27031.14,1.07,1.14,53.51,26.27
min,0.000000e+00,5.250000e+02,0.000000e+00,0.00,0.00,0.00,0.00,-288.00,1776.00
25%,5.850000e+05,5.751750e+05,5.750000e+05,0.12,1247.00,3.00,2.00,8.00,1960.00
50%,8.250000e+05,8.150000e+05,8.200000e+05,0.17,1641.00,3.00,2.00,19.00,1979.00
75%,1.299000e+06,1.295000e+06,1.300000e+06,0.27,2217.00,4.00,3.00,48.00,1999.00
max,1.390000e+09,1.375000e+08,9.895000e+08,7810698.36,17021321.00,45.00,175.00,12430.00,2026.00



### Price Fields

| Field | Count | Median | Mean | Min | Max | Flag |
|-------|-------|--------|------|-----|-----|------|
| `OriginalListPrice` | 396,496 | $825,000 | $1.22M | $0 | $1.39B | 🔴 Outliers |
| `ListPrice` | 397,210 | $815,000 | $1.14M | $525 | $137.5M | 🔴 Outliers |
| `ClosePrice` | 397,208 | $820,000 | $1.19M | $0 | $989.5M | 🔴 Outliers |

> Unlike the listing dataset where `ClosePrice` covers only ~26% of records, **ClosePrice is fully populated here** (~100%). Use this dataset for all sold price analyses. The mean is ~$370K above the median — driven by high-end luxury sales.

---

### Property Attribute Fields

| Field | Count | Median | Mean | Min | Max | Flag |
|-------|-------|--------|------|-----|-----|------|
| `LivingArea` | 396,982 | 1,641 sqft | 1,904 sqft | 0 | 17,021,321 sqft | 🔴 Outliers |
| `LotSizeAcres` | 365,892 | 0.17 ac | 68.4 ac | 0 | 7,810,698 ac | 🔴 Outliers |
| `BedroomsTotal` | 397,199 | 3 | 3.20 | 0 | 45 | 🟠 Review |
| `BathroomsTotalInteger` | 397,141 | 2 | 2.53 | 0 | 175 | 🔴 Outliers |
| `DaysOnMarket` | 397,210 | 19 | 37.3 | −288 | 12,430 | 🔴 Outliers |
| `YearBuilt` | 396,854 | 1979 | 1978.6 | 1776 | 2026 | 🟠 Review |

**Outlier explanations:**

| Value | Likely Cause |
|-------|-------------|
| `OriginalListPrice` / `ClosePrice` = $0 | Missing or placeholder — remove |
| `LivingArea` = 17M sqft | Unit mismatch or data entry error |
| `LotSizeAcres` = 7.81M ac | Unit error or erroneous entry |
| `DaysOnMarket` = −288 | Calculation error — must be removed |
| `DaysOnMarket` = 12,430 | Clear data error — must be removed |
| `BathroomsTotalInteger` = 175 | Clear data entry error |
| `BedroomsTotal` = 45 | Possible commercial misclassification |
| `YearBuilt` = 1776 | Likely a placeholder value |

---

## 5. Intern Questions — Answered

### Q1. What is the Residential vs. other property type share?



In [67]:
# PropertySubType distribution
subtype_dist = df['PropertySubType'].value_counts(dropna=False)
subtype_pct = df['PropertySubType'].value_counts(normalize=True, dropna=False) * 100

subtype_summary = pd.DataFrame({
    'count': subtype_dist,
    'share_%': subtype_pct.round(2)
})

print(subtype_summary)

                        count  share_%
PropertySubType                       
SingleFamilyResidence  297065    74.79
Condominium             65748    16.55
Townhouse               23324     5.87
ManufacturedOnLand       5120     1.29
Duplex                   2157     0.54
StockCooperative         1601     0.40
NaN                       776     0.20
Cabin                     452     0.11
Triplex                   324     0.08
MixedUse                  193     0.05
Quadruplex                133     0.03
BoatSlip                   81     0.02
OwnYourOwn                 57     0.01
MobileHome                 49     0.01
ManufacturedHome           39     0.01
Loft                       26     0.01
Timeshare                  18     0.00
CoOwnership                16     0.00
Farm                       15     0.00
Studio                     10     0.00
DeededParking               6     0.00


> The dataset is **100% Residential** and **100% Closed** — both filters applied upstream.  
> Breakdown by `PropertySubType`:

| Property Sub Type | Count | Share |
|-------------------|-------|-------|
| `SingleFamilyResidence` | 297,065 | ~75% |
| Condominium & others | ~100,145 | ~25% |

---

### Q2. What are the median and average close prices?

> `ClosePrice` is populated for **397,208 records** (~100% of total).

| Metric | Value |
|--------|-------|
| **Median Close Price** | **$820,000** |
| **Mean Close Price** | **$1,190,000** |
| Gap | ~$370,000 (mean is ~45% above median) |

> The gap is driven by high-end luxury sales pulling the mean upward. For market reporting, **median is the preferred benchmark**.

---

### Q3. What is the price ratio (ClosePrice / ListPrice)?

In [68]:

sold = df.dropna(subset=['ClosePrice', 'ListPrice']).copy()
sold['price_ratio'] = sold['ClosePrice'] / sold['ListPrice']

above = (sold['price_ratio'] > 1).sum()
at    = (sold['price_ratio'] == 1).sum()
below = (sold['price_ratio'] < 1).sum()
total = len(sold)

print(f"Above list price: {above:,} ({above/total*100:.1f}%)")
print(f"At list price:    {at:,} ({at/total*100:.1f}%)")
print(f"Below list price: {below:,} ({below/total*100:.1f}%)")
print(f"Median ratio:     {sold['price_ratio'].median():.4f}")


Above list price: 159,344 (40.1%)
At list price:    68,873 (17.3%)
Below list price: 168,991 (42.5%)
Median ratio:     1.0000


### Q4. What percentage of homes sold above vs. below list price?

In [69]:
# ── Q4. Sold Above vs. Below List Price ─────────────────────────────────────
sold = df.dropna(subset=['ClosePrice', 'ListPrice']).copy()
sold['price_ratio'] = sold['ClosePrice'] / sold['ListPrice']

above = (sold['price_ratio'] > 1).sum()
at    = (sold['price_ratio'] == 1).sum()
below = (sold['price_ratio'] < 1).sum()
total = len(sold)

print("=== Q4. Sold Price vs. List Price ===")
print(f"Total closed listings:  {total:,}")
print(f"Above list price:       {above:,} ({above/total*100:.1f}%)")
print(f"At list price:          {at:,} ({at/total*100:.1f}%)")
print(f"Below list price:       {below:,} ({below/total*100:.1f}%)")
print(f"Median Close/List ratio:{sold['price_ratio'].median():.4f}")
print(f"Mean Close/List ratio:  {sold['price_ratio'].mean():.4f}")

=== Q4. Sold Price vs. List Price ===
Total closed listings:  397,208
Above list price:       159,344 (40.1%)
At list price:          68,873 (17.3%)
Below list price:       168,991 (42.5%)
Median Close/List ratio:1.0000
Mean Close/List ratio:  1.0780


### Q5. Are there any apparent date consistency issues (e.g., close date before listing date)?

In [70]:
# ── Q5. Date Consistency Issues ─────────────────────────────────────────────
df['CloseDate_dt']    = pd.to_datetime(df['CloseDate'], errors='coerce')
df['ListingDate_dt']  = pd.to_datetime(df['ListingContractDate'], errors='coerce')
df['PurchaseDate_dt'] = pd.to_datetime(df['PurchaseContractDate'], errors='coerce')

# Close before listing
c_before_l = df.dropna(subset=['CloseDate_dt', 'ListingDate_dt'])
close_before_list = (c_before_l['CloseDate_dt'] < c_before_l['ListingDate_dt']).sum()

# Purchase before listing
p_before_l = df.dropna(subset=['PurchaseDate_dt', 'ListingDate_dt'])
purchase_before_list = (p_before_l['PurchaseDate_dt'] < p_before_l['ListingDate_dt']).sum()

# Close before purchase
c_before_p = df.dropna(subset=['CloseDate_dt', 'PurchaseDate_dt'])
close_before_purchase = (c_before_p['CloseDate_dt'] < c_before_p['PurchaseDate_dt']).sum()

print("=== Q5. Date Consistency Issues ===")
print(f"Close before listing date:    {close_before_list:,}")
print(f"Purchase before listing date: {purchase_before_list:,}")
print(f"Close before purchase date:   {close_before_purchase:,}")

=== Q5. Date Consistency Issues ===
Close before listing date:    58
Purchase before listing date: 260
Close before purchase date:   240



### Q6.Which counties have the highest median prices?

In [71]:
# ── Q6. County Median Prices ─────────────────────────────────────────────────
county = (
    df.dropna(subset=['ClosePrice'])
    .groupby('CountyOrParish')
    .agg(
        median_close_price=('ClosePrice', 'median'),
        median_list_price=('ListPrice', 'median'),
        sold_count=('ClosePrice', 'count')
    )
    .sort_values('median_close_price', ascending=False)
    .reset_index()
)

county['median_close_price'] = county['median_close_price'].apply(lambda x: f"${x:,.0f}")
county['median_list_price']  = county['median_list_price'].apply(lambda x: f"${x:,.0f}")

print("=== Q6. Median Close Price by County (Top 15) ===")
print(county.head(15).to_string(index=False))

=== Q6. Median Close Price by County (Top 15) ===
CountyOrParish median_close_price median_list_price  sold_count
     Del Norte         $2,485,000        $2,485,000           1
  Other County         $2,462,500          $332,000           4
     San Mateo         $1,700,000        $1,599,000        7004
   Santa Clara         $1,600,000        $1,499,000       17855
    Santa Cruz         $1,200,000        $1,199,000        2900
 San Francisco         $1,180,000        $1,088,000         921
        Orange         $1,175,000        $1,170,000       44711
         Marin         $1,170,000        $1,195,000         139
       Alameda         $1,135,000        $1,000,000       18729
        Alpine         $1,100,000        $1,295,000           1
          Mono           $915,000          $975,000          13
      Monterey           $907,000          $907,000        3842
   Los Angeles           $900,000          $899,000       98730
     San Diego           $896,000          $899,000   